src/test/ensemble_flow_custom_node_tests.cpp

```diff
@@ -0,0 +1,974 @@
+
+        for (size_t i = 0; i < actual.size(); i++) {
+            EXPECT_NEAR(actual[i], data[i], 0.001);
+        }
+    }
+
+    template <typename T>
+    static NodeLibrary createLibraryMock() {
+        return NodeLibrary{
+            T::execute,
+            T::releaseBuffer,
+            T::releaseTensors};
+    }
+
+    PredictRequest request;
+    PredictResponse response;
+
+    NodeLibrary library;
+
+    const std::string customNodeName = "add_sub_node";
+    const std::string libraryName = "add_sub_lib";
+    const std::string libraryPath = "/ovms/bazel-bin/src/lib_node_add_sub.so";
+    const std::string customNodeInputName = "input_numbers";
+    const std::string customNodeOutputName = "output_numbers";
+    const std::string pipelineInputName = "pipeline_input";
+    const std::string pipelineOutputName = "pipeline_output";
+};
+
+TEST_F(EnsembleFlowCustomNodePipelineExecutionTest, AddSubCustomNode) {
+    // Most basic configuration, just process single add-sub custom node pipeline request
+    // input  add-sub  output
+    //  O------->O------->O
+    const std::vector<float> inputValues{3.2, 5.7, -2.4};
+    this->prepareRequest(inputValues);
+
+    const float addValue = 2.5;
+    const float subValue = 4.8;
+
+    auto input_node = std::make_unique<EntryNode>(&request);
+    auto output_node = std::make_unique<ExitNode>(&response);
+    auto custom_node = std::make_unique<CustomNode>(customNodeName, library,
+        parameters_t{
+            {"add_value", std::to_string(addValue)},
+            {"sub_value", std::to_string(subValue)}});
+
+    Pipeline pipeline(*input_node, *output_node);
+    pipeline.connect(*input_node, *custom_node, {{pipelineInputName, customNodeInputName}});
+    pipeline.connect(*custom_node, *output_node, {{customNodeOutputName, pipelineOutputName}});
+
+    pipeline.push(std::move(input_node));
+    pipeline.push(std::move(custom_node));
+    pipeline.push(std::move(output_node));
+
+    ASSERT_EQ(pipeline.execute(), StatusCode::OK);
+    ASSERT_EQ(response.outputs().size(), 1);
+
+    this->checkResponse<float>(inputValues, [addValue, subValue](float value) -> float {
+        return value + addValue - subValue;
+    });
+}
+
+TEST_F(EnsembleFlowCustomNodePipelineExecutionTest, SeriesOfCustomNodes) {
+    constexpr int N = 100;
+    constexpr int PARAMETERS = 2;
+    static_assert(PARAMETERS > 0);
+    static_assert(N > PARAMETERS);
+    static_assert((N % PARAMETERS) == 0);
+    // input      add-sub x N      output
+    //  O------->O->O...O->O------->O
+
+    const std::vector<float> inputValues{3.2, 5.7, -2.4};
+    this->prepareRequest(inputValues);
+
+    const std::array<float, PARAMETERS> addValue{1.5, -2.4};
+    const std::array<float, PARAMETERS> subValue{-5.1, 1.9};
+
+    auto input_node = std::make_unique<EntryNode>(&request);
+    auto output_node = std::make_unique<ExitNode>(&response);
+
+    std::unique_ptr<CustomNode> custom_nodes[N];
+    for (int i = 0; i < N; i++) {
```

Comment:
Wouldn't it be better to use odd N to at least know that first addValue was used different number of times than the second one? Otherwise I am not sure if complicating this test by alternating is worthy the complication.

Reply:
Then this test's constant variables have no sense - all the scripting around test configuration is to allow simple change by playing N & PARAMETERS constants and not needing to touch `checkResults` and asserts at all. When using odd number, somebody will need to change the `checkResults` aswell.

Reply:
I think its worthy to have different parameters for nodes to check if one library can handle multiple purposes just by changing parameters.

Reply:
This test makes sure that shared library does not somehow cache parameters